python/dgl/heterograph.py

```diff
@@ -0,0 +1,1387 @@
+        edges : valid edges type, dict, optional
+            Edges on which to group and apply ``func``. See :func:`send` for valid
+            edges type. Default is all the edges.
+
+            If ``func`` is not a dict, then ``edges`` must not be a dict.
+            If ``func`` is a dict, then ``edges`` must either be
+            * ALL: for computing on all edges with the given types in ``func``.
+            * a dict of int, iterable of int, or tensors, with the same keys
+              as ``func``, indicating the edges to be updated for each type.
+        inplace: bool, optional
+            If True, update will be done in place, but autograd will break.
+        """
+        pass
+
+    # TODO: REVIEW
+    def send(self, edges=ALL, message_func=None):
+        """Send messages along the given edges with the same edge type.
+
+        ``edges`` can be any of the following types:
+
+        * ``int`` : Specify one edge using its edge id.
+        * ``pair of int`` : Specify one edge using its endpoints.
+        * ``int iterable`` / ``tensor`` : Specify multiple edges using their edge ids.
+        * ``pair of int iterable`` / ``pair of tensors`` :
+          Specify multiple edges using their endpoints.
+
+        The UDF returns messages on the edges and can be later fetched in
+        the destination node's ``mailbox``. Receiving will consume the messages.
+        See :func:`recv` for example.
+
+        If multiple ``send`` are triggered on the same edge without ``recv``. Messages
+        generated by the later ``send`` will overwrite previous messages.
+
+        Parameters
+        ----------
+        edges : valid edges type, dict, optional
+            Edges on which to apply ``message_func``. Default is sending along all
+            the edges.
+
+            If ``message_func`` is not a dict, then ``edges`` must not be a dict.
+            If ``message_func`` is a dict, then ``edges`` must either be
+            * ALL: for computing on all edges with the given types in
+              ``message_func``.
+            * a dict of int, iterable of int, or tensors, with the same keys
+              as ``message_func``, indicating the edges to be updated for each
+              type.
+        message_func : callable, dict[etype, callable]
+            Message function on the edges. The function should be
+            an :mod:`Edge UDF <dgl.udf>`.
+
+            If a dict is provided, the functions will be applied according to
+            edge type.
+            The edge type is characterized by a triplet of source type name,
+            destination type name, and edge type name.
+            If the graph has more than one edge type and ``message_func`` is
+            not a dict, it will throw an error.
+
+        Notes
+        -----
+        On multigraphs, if :math:`u` and :math:`v` are specified, then the messages will be sent
+        along all edges between :math:`u` and :math:`v`.
```

Comment:
How do we specify `u` and `v` if they have different node types? Do we have to use global ID?